### DISCUSSION TITLE
Why/how does the object mapper allow source `array` but not `stdClass`?

### DISCUSSION THREAD

**Discussion by ToshY** [rekalogika/mapper#18] [Q&A] [answered]
I was using the following snippet for the mapper:

```php
$this->mapper->map(['finished' => true], BookDto::class);
```

Now I noticed my PHPstan was not happy as I was passing an `array` as source to the mapper, while it should've accepted an `object` according to the `MapperInterface`. Funnily enough it works, so I'm not complaining, but it left me in a state of confusing and I'm now just wondering why it works though 🤔

So is it designed to handle `array` as source or am I accidently abusing the mapper by doing something like this?

---

Now I was trying to prevent these PHPstan warnings and I thought maybe explicitely casting the array to an object would do the trick:

```php
$this->mapper->map((object) ['finished' => true], BookDto::class);
```

But the mapper doesn't like this as it throws an `InternalClassUnsupportedException` with the message: `Trying to map an internal class "stdClass" which is not supported. Mapping path: "(root)".` (due to `$sourceReflection->isInternal()` in the `ObjectToObjectMetadataFactory::87`).

I was looking through the docs to something related to this, but couldn't seem to find it, so could you point me in the right direction, or maybe suggest an alternative on how to solve this problem?

**Version: `1.0.0`**

**Comment by priyadi** [comment, marked as answer]
> ```php
> $this->mapper->map(['finished' => true], BookDto::class);
> ```

The code works because there is a bug in the implementation that it accepts mixed, not object, unlike the interface. Liskov allows it, so static analysis did not catch the bug. I'll fix the bug later.

Internally, the engine [can handle array to object mapping](https://rekalogika.dev/mapper/object-array). It just forwards the job to Symfony Serializer. So your code can work.

PHPstan does not like it because it sees `$this->mapper` as a `MapperInterface`, not the concrete `Mapper`. But PHP doesn't care, it just runs the code.

I blacklisted internal classes because it almost always results in an error with an obscure error message. `stdClass` should be specifically supported, though. I'm adding it to my to-do list.

**Reply by ToshY** [reply]
Thanks for the detailed explanation 👍
